src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBiomartRetriever.java

```diff
@@ -0,0 +1,232 @@
+     * @param fileSuffix - String that will be used to make filename (either _uniprot or _microarray) that will hold associated data.
+     * @throws IOException - Thrown when unable to write data to file.
+     */
+    private void queryBiomartAndStoreData(String biomartSpeciesName, String biomartQueryFilePath, String biomartDataType, String fileSuffix) throws IOException {
+
+        Set<String> biomartResponseLines = new HashSet<>();
+        try {
+            logger.info("Retrieving data associated with microarray type " + biomartDataType);
+            biomartResponseLines = queryBiomart(getBiomartIdentifierQuery(biomartQueryFilePath, biomartSpeciesName, biomartDataType), 0);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+
+        String biomartFilename = this.destination + biomartSpeciesName + fileSuffix;
+        storeBiomartData(biomartFilename, biomartResponseLines);
+
+    }
+
+    /**
+     * Write response from BioMart to file in the format 'biomartSpeciesName_microarray' or 'biomartSpeciesName_uniprot'.
+     * @param biomartFilename - String, Data from BioMart is saved to this filename.
+     * @param biomartResponseLines - Set<String>, All lines of response from BioMart.
+     * @throws IOException - Thrown when unable to write data to file.
+     */
+    private void storeBiomartData(String biomartFilename, Set<String> biomartResponseLines) throws IOException {
+
+        Set<String> biomartIdentifierLines = new HashSet<>();
+        for (String biomartResponseLine : biomartResponseLines) {
+            List<String> tabSplit = Arrays.asList(biomartResponseLine.split("\t"));
+            if (lineContainsIdentifier(tabSplit)) {
+                biomartIdentifierLines.add(biomartResponseLine + "\n");
+            }
+        }
+
+        // Write data to file.
+        for (String biomartIdentifierLine : biomartIdentifierLines) {
+            Files.write(
+                    Paths.get(biomartFilename),
+                    biomartIdentifierLine.getBytes(),
+                    StandardOpenOption.APPEND, StandardOpenOption.CREATE
+            );
+
+        }
+    }
+
+    // Checks that the 4th column contains identifiers (either Microarray or UniProt ).
+    private boolean lineContainsIdentifier(List<String> tabSplit) {
+        final int IDENTIFIER_COLUMN = 3;
+        return tabSplit.size() > IDENTIFIER_COLUMN && !tabSplit.get(IDENTIFIER_COLUMN).trim().isEmpty();
+    }
+
+
+    /**
+     * This method queries BioMart using either a URL (see variable 'microarrayTypesBaseQuery) or an XML (see biomart-query.xml in resources) query.
+     * It will retry up to 5 times if errors are returned from BioMart instead of data.
+     * There are multiple cases (Frog, Both Yeasts, P. falciparum and D. discoideum) where the data does not exist in BioMart.
+     * If/When this exception is thrown for this species, it can be ignored. Check the logs from previous runs to confirm.
```

Comment:
Is this a Java exception that's thrown?  If so, it'd be good to mention the specific type.

Reply:
No, that was from when it used to be a general Exception. This actually describes the BioMartQueryException. I moved this line to that description.